File: sdk/typescript/smartc-simple.ts
/**
 * SmaRTC TypeScript SDK - Wrapper Simplifié avec types
 * 
 * @packageDocumentation
 * @module smartc-sdk
 */

// ==========================================
// 🔧 Configuration & Types
// ==========================================

export interface SmaRTCConfig {
  apiUrl?: string;
  signalServerUrl?: string;
  stunServers?: string[];
  turnServers?: TurnServer[];
  timeout?: number;
}

export interface TurnServer {
  urls: string | string[];
  username?: string;
  credential?: string;
}

export interface User {
  id: number;
  username: string;
  role: string;
}

export interface Session {
  id: number;
  name: string;
  description?: string;
  creatorId: number;
  createdAt: string;
}

export interface ICEServer {
  urls: string | string[];
  username?: string;
  credential?: string;
}

export interface LoginResponse {
  token: string;
  user?: User;
}

export interface RegisterRequest {
  username: string;
  password: string;
  role?: string;
}

export interface LoginRequest {
  username: string;
  password: string;
}

export interface CreateSessionRequest {
  name: string;
  description?: string;
}

// ==========================================
// ❌ Custom Errors
// ==========================================

export class SmaRTCError extends Error {
  constructor(
    message: string,
    public readonly original?: Error
  ) {
    super(message);
    this.name = 'SmaRTCError';
  }
}

export class AuthenticationError extends SmaRTCError {
  constructor(message: string = 'Identifiants incorrects', original?: Error) {
    super(message, original);
    this.name = 'AuthenticationError';
  }
}

export class SessionNotFoundError extends SmaRTCError {
  constructor(message: string = 'Cet appel n\'existe pas', original?: Error) {
    super(message, original);
    this.name = 'SessionNotFoundError';
  }
}

export class NetworkError extends SmaRTCError {
  constructor(message: string = 'Problème de connexion', original?: Error) {
    super(message, original);
    this.name = 'NetworkError';
  }
}

// ==========================================
// 🚀 SmaRTC Client Simplifié
// ==========================================

export class SmaRTCSimple {
  private config: Required<SmaRTCConfig>;
  private token: string | null = null;
  private username: string | null = null;
  private currentSessionId: number | null = null;

  constructor(config: SmaRTCConfig = {}) {
    this.config = {
      apiUrl: config.apiUrl || 'http://localhost:8080',
      signalServerUrl: config.signalServerUrl || 'http://localhost:5001/signalhub',
      stunServers: config.stunServers || ['stun:stun.l.google.com:19302'],
      turnServers: config.turnServers || [],
      timeout: config.timeout || 30000,
    };
  }

  // ==========================================
  // 🔒 Private Helpers
  // ==========================================

  private getHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };

    if (this.token) {
      headers['Authorization'] = `Bearer ${this.token}`;
    }

    return headers;
  }

  private async request<T = any>(
    method: string,
    endpoint: string,
    body?: any
  ): Promise<T> {
    const url = `${this.config.apiUrl}${endpoint}`;
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.config.timeout);

    try {
      const response = await fetch(url, {
        method,
        headers: this.getHeaders(),
        body: body ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });

      clearTimeout(timeout);

      if (response.status === 401) {
        throw new AuthenticationError();
      }

      if (response.status === 404) {
        throw new SessionNotFoundError();
      }

      if (response.status === 409) {
        throw new SmaRTCError('Ce nom d\'utilisateur existe déjà');
      }

      if (!response.ok) {
        const text = await response.text();
        throw new SmaRTCError(`Erreur HTTP ${response.status}: ${text}`);
      }

      if (response.status === 204) {
        return {} as T;
      }

      return await response.json();
    } catch (error) {
      clearTimeout(timeout);

      if (error instanceof SmaRTCError) {
        throw error;
      }

      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          throw new NetworkError('Timeout de connexion', error);
        }
        throw new NetworkError(`Problème de connexion: ${error.message}`, error);
      }

      throw new SmaRTCError('Erreur inconnue', error as Error);
    }
  }

  // ==========================================
  // 🔐 Authentification
  // ==========================================

  /**
   * Se connecter avec username et password
   * @param username - Nom d'utilisateur
   * @param password - Mot de passe
   * @returns Promise<true> si succès
   * @throws {AuthenticationError} Si identifiants incorrects
   * @throws {NetworkError} Si problème de connexion
   */
  async login(username: string, password: string): Promise<boolean> {
    try {
      const data = await this.request<LoginResponse>('POST', '/api/auth/login', {
        username,
        password,
      });

      this.token = data.token;
      this.username = username;

      return true;
    } catch (error) {
      if (error instanceof SmaRTCError) {
        throw error;
      }
      throw new SmaRTCError('Erreur de connexion', error as Error);
    }
  }

  /**
   * Créer un nouveau compte
   * @param username - Nom d'utilisateur
   * @param password - Mot de passe
   * @param role - Rôle (User ou Admin)
   * @returns Promise<true> si succès
   * @throws {SmaRTCError} Si le nom d'utilisateur existe déjà
   */
  async register(username: string, password: string, role: string = 'User'): Promise<boolean> {
    try {
      await this.request('POST', '/api/auth/register', {
        username,
        password,
        role,
      });

      return true;
    } catch (error) {
      if (error instanceof SmaRTCError) {
        throw error;
      }
      throw new SmaRTCError('Erreur d\'inscription', error as Error);
    }
  }

  /**
   * Se déconnecter et nettoyer la session
   */
  async logout(): Promise<void> {
    if (this.currentSessionId) {
      await this.endCall();
    }

    this.token = null;
    this.username = null;
  }

  /**
   * Vérifie si l'utilisateur est connecté
   */
  get isLoggedIn(): boolean {
    return this.token !== null;
  }

  /**
   * Récupère le nom d'utilisateur actuel
   */
  get currentUsername(): string | null {
    return this.username;
  }

  // ==========================================
  // 📹 Appels vidéo
  // ==========================================

  /**
   * Démarre un appel vidéo (crée une session)
   * @param name - Nom de l'appel
   * @param description - Description optionnelle
   * @returns Promise<number> ID de la session créée
   * @throws {SmaRTCError} Si impossible de créer l'appel
   */
  async startCall(name: string, description?: string): Promise<number> {
    try {
      const session = await this.request<Session>('POST', '/api/session', {
        name,
        description,
      });

      this.currentSessionId = session.id;
      return session.id;
    } catch (error) {
      if (error instanceof SmaRTCError) {
        throw error;
      }
      throw new SmaRTCError('Impossible de créer l\'appel', error as Error);
    }
  }

  /**
   * Rejoindre un appel existant
   * @param sessionId - ID de la session à rejoindre
   * @throws {SessionNotFoundError} Si la session n'existe pas
   */
  async joinCall(sessionId: number): Promise<void> {
    try {
      await this.request('GET', `/api/session/${sessionId}`);
      this.currentSessionId = sessionId;
    } catch (error) {
      if (error instanceof SessionNotFoundError) {
        throw error;
      }
      throw new SmaRTCError('Erreur lors de la connexion', error as Error);
    }
  }

  /**
   * Termine l'appel en cours
   */
  async endCall(): Promise<void> {
    if (!this.currentSessionId) {
      return;
    }

    try {
      // Optionnel: supprimer la session si créateur
      // await this.request('DELETE', `/api/session/${this.currentSessionId}`);
      this.currentSessionId = null;
    } catch {
      // Ignorer les erreurs lors de la fermeture
    }
  }

  // ==========================================
  // 📋 Sessions disponibles
  // ==========================================

  /**
   * Liste tous les appels disponibles
   * @returns Promise<Session[]> Liste des sessions
   */
  async getAvailableCalls(): Promise<Session[]> {
    try {
      return await this.request<Session[]>('GET', '/api/session');
    } catch (error) {
      if (error instanceof SmaRTCError) {
        throw error;
      }
      throw new SmaRTCError('Erreur lors de la récupération', error as Error);
    }
  }

  /**
   * Récupère les détails d'un appel
   * @param sessionId - ID de la session
   * @returns Promise<Session> Détails de la session
   */
  async getCallDetails(sessionId: number): Promise<Session> {
    try {
      return await this.request<Session>('GET', `/api/session/${sessionId}`);
    } catch (error) {
      if (error instanceof SessionNotFoundError) {
        throw new SmaRTCError('Appel introuvable');
      }
      if (error instanceof SmaRTCError) {
        throw error;
      }
      throw new SmaRTCError('Erreur', error as Error);
    }
  }

  // ==========================================
  // 🌐 Configuration réseau
  // ==========================================

  /**
   * Récupère les serveurs ICE (STUN/TURN)
   * @returns Promise<ICEServer[]> Liste des serveurs ICE avec fallback
   */
  async getIceServers(): Promise<ICEServer[]> {
    try {
      const data = await this.request<{ iceServers: ICEServer[] }>('GET', '/api/webrtc/ice');
      return data.iceServers || [];
    } catch {
      // Fallback sur Google STUN
      return this.config.stunServers.map(urls => ({ urls }));
    }
  }
}

// Export par défaut pour import facile
export default SmaRTCSimple;

File: SmaRTC-core/sdk/kotlin/SmaRTCSimple.kt
/**
 * SmaRTC Kotlin SDK - Wrapper Simplifié pour Android
 * 
 * @author DeLTa-X Tunisia
 * @version 1.0.0
 */

package tn.smartc.sdk

import kotlinx.coroutines.*
import kotlinx.serialization.*
import kotlinx.serialization.json.Json
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

// ==========================================
// 🔧 Configuration & Models
// ==========================================

@Serializable
data class SmaRTCConfig(
    val apiUrl: String = "http://10.0.2.2:8080", // Android emulator localhost
    val signalServerUrl: String = "http://10.0.2.2:5001/signalhub",
    val stunServers: List<String> = listOf("stun:stun.l.google.com:19302"),
    val turnServers: List<TurnServer> = emptyList(),
    val timeout: Long = 30000
)

@Serializable
data class TurnServer(
    val urls: String,
    val username: String? = null,
    val credential: String? = null
)

@Serializable
data class User(
    val id: Int,
    val username: String,
    val role: String
)

@Serializable
data class Session(
    val id: Int,
    val name: String,
    val description: String? = null,
    val creatorId: Int,
    val createdAt: String
)

@Serializable
data class LoginResponse(
    val token: String
)

@Serializable
data class LoginRequest(
    val username: String,
    val password: String
)

@Serializable
data class RegisterRequest(
    val username: String,
    val password: String,
    val role: String = "User"
)

@Serializable
data class CreateSessionRequest(
    val name: String,
    val description: String? = null
)

@Serializable
data class ICEServerResponse(
    val iceServers: List<ICEServer>
)

@Serializable
data class ICEServer(
    val urls: String
)

// ==========================================
// ❌ Custom Exceptions
// ==========================================

sealed class SmaRTCException(message: String, cause: Throwable? = null) : Exception(message, cause) {
    class AuthenticationError(message: String = "Identifiants incorrects", cause: Throwable? = null) : 
        SmaRTCException(message, cause)
    
    class SessionNotFoundError(message: String = "Cet appel n'existe pas", cause: Throwable? = null) : 
        SmaRTCException(message, cause)
    
    class NetworkError(message: String = "Problème de connexion", cause: Throwable? = null) : 
        SmaRTCException(message, cause)
    
    class GenericError(message: String, cause: Throwable? = null) : 
        SmaRTCException(message, cause)
}

// ==========================================
// 🚀 SmaRTC Client Simplifié
// ==========================================

class SmaRTCSimple(private val config: SmaRTCConfig = SmaRTCConfig()) {
    
    private val json = Json { ignoreUnknownKeys = true }
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(config.timeout, TimeUnit.MILLISECONDS)
        .readTimeout(config.timeout, TimeUnit.MILLISECONDS)
        .writeTimeout(config.timeout, TimeUnit.MILLISECONDS)
        .build()
    
    private var token: String? = null
    private var username: String? = null
    private var currentSessionId: Int? = null
    
    companion object {
        private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
    }
    
    // ==========================================
    // 🔒 Private Helpers
    // ==========================================
    
    private fun getHeaders(): Headers.Builder {
        val builder = Headers.Builder()
            .add("Content-Type", "application/json")
        
        token?.let {
            builder.add("Authorization", "Bearer $it")
        }
        
        return builder
    }
    
    private suspend fun <T> request(
        method: String,
        endpoint: String,
        body: Any? = null,
        parser: (String) -> T
    ): T = withContext(Dispatchers.IO) {
        val url = "${config.apiUrl}$endpoint"
        
        val requestBuilder = Request.Builder()
            .url(url)
            .headers(getHeaders().build())
        
        when (method) {
            "GET" -> requestBuilder.get()
            "POST" -> {
                val requestBody = body?.let {
                    json.encodeToString(it).toRequestBody(JSON_MEDIA_TYPE)
                } ?: "".toRequestBody(JSON_MEDIA_TYPE)
                requestBuilder.post(requestBody)
            }
            "PUT" -> {
                val requestBody = body?.let {
                    json.encodeToString(it).toRequestBody(JSON_MEDIA_TYPE)
                } ?: "".toRequestBody(JSON_MEDIA_TYPE)
                requestBuilder.put(requestBody)
            }
            "DELETE" -> requestBuilder.delete()
        }
        
        try {
            val response = client.newCall(requestBuilder.build()).execute()
            
            when (response.code) {
                401 -> throw SmaRTCException.AuthenticationError()
                404 -> throw SmaRTCException.SessionNotFoundError()
                409 -> throw SmaRTCException.GenericError("Ce nom d'utilisateur existe déjà")
                in 400..599 -> throw SmaRTCException.GenericError("Erreur HTTP ${response.code}")
                204 -> return@withContext parser("{}")
            }
            
            val responseBody = response.body?.string() 
                ?: throw SmaRTCException.NetworkError("Réponse vide")
            
            parser(responseBody)
        } catch (e: IOException) {
            throw SmaRTCException.NetworkError("Problème de connexion: ${e.message}", e)
        } catch (e: SmaRTCException) {
            throw e
        } catch (e: Exception) {
            throw SmaRTCException.GenericError("Erreur inattendue: ${e.message}", e)
        }
    }
    
    // ==========================================
    // 🔐 Authentification
    // ==========================================
    
    /**
     * Se connecter avec username et password
     * @param username Nom d'utilisateur
     * @param password Mot de passe
     * @return true si succès
     * @throws SmaRTCException.AuthenticationError Si identifiants incorrects
     * @throws SmaRTCException.NetworkError Si problème de connexion
     */
    suspend fun login(username: String, password: String): Boolean {
        return try {
            val response = request(
                "POST",
                "/api/auth/login",
                LoginRequest(username, password)
            ) { json.decodeFromString<LoginResponse>(it) }
            
            this.token = response.token
            this.username = username
            
            true
        } catch (e: SmaRTCException) {
            throw e
        } catch (e: Exception) {
            throw SmaRTCException.GenericError("Erreur de connexion", e)
        }
    }
    
    /**
     * Créer un nouveau compte
     * @param username Nom d'utilisateur
     * @param password Mot de passe
     * @param role Rôle (User ou Admin)
     * @return true si succès
     * @throws SmaRTCException.GenericError Si le nom d'utilisateur existe déjà
     */
    suspend fun register(username: String, password: String, role: String = "User"): Boolean {
        return try {
            request(
                "POST",
                "/api/auth/register",
                RegisterRequest(username, password, role)
            ) { Unit }
            
            true
        } catch (e: SmaRTCException) {
            throw e
        } catch (e: Exception) {
            throw SmaRTCException.GenericError("Erreur d'inscription", e)
        }
    }
    
    /**
     * Se déconnecter et nettoyer la session
     */
    suspend fun logout() {
        currentSessionId?.let { endCall() }
        token = null
        username = null
    }
    
    /**
     * Vérifie si l'utilisateur est connecté
     */
    val isLoggedIn: Boolean
        get() = token != null
    
    /**
     * Récupère le nom d'utilisateur actuel
     */
    val currentUsername: String?
        get() = username
    
    // ==========================================
    // 📹 Appels vidéo
    // ==========================================
    
    /**
     * Démarre un appel vidéo (crée une session)
     * @param name Nom de l'appel
     * @param description Description optionnelle
     * @return ID de la session créée
     * @throws SmaRTCException.GenericError Si impossible de créer l'appel
     */
    suspend fun startCall(name: String, description: String? = null): Int {
        return try {
            val session = request(
                "POST",
                "/api/session",
                CreateSessionRequest(name, description)
            ) { json.decodeFromString<Session>(it) }
            
            currentSessionId = session.id
            session.id
        } catch (e: SmaRTCException) {
            throw e
        } catch (e: Exception) {
            throw SmaRTCException.GenericError("Impossible de créer l'appel", e)
        }
    }
    
    /**
     * Rejoindre un appel existant
     * @param sessionId ID de la session à rejoindre
     * @throws SmaRTCException.SessionNotFoundError Si la session n'existe pas
     */
    suspend fun joinCall(sessionId: Int) {
        try {
            request("GET", "/api/session/$sessionId") { 
                json.decodeFromString<Session>(it)
            }
            currentSessionId = sessionId
        } catch (e: SmaRTCException.SessionNotFoundError) {
            throw e
        } catch (e: Exception) {
            throw SmaRTCException.GenericError("Erreur lors de la connexion", e)
        }
    }
    
    /**
     * Termine l'appel en cours
     */
    suspend fun endCall() {
        currentSessionId?.let {
            try {
                // Optionnel: supprimer la session
                // request<Unit>("DELETE", "/api/session/$it") { }
                currentSessionId = null
            } catch (e: Exception) {
                // Ignorer les erreurs
            }
        }
    }
    
    // ==========================================
    // 📋 Sessions disponibles
    // ==========================================
    
    /**
     * Liste tous les appels disponibles
     * @return Liste des sessions
     */
    suspend fun getAvailableCalls(): List<Session> {
        return try {
            request("GET", "/api/session") {
                json.decodeFromString<List<Session>>(it)
            }
        } catch (e: SmaRTCException) {
            throw e
        } catch (e: Exception) {
            throw SmaRTCException.GenericError("Erreur lors de la récupération", e)
        }
    }
    
    /**
     * Récupère les détails d'un appel
     * @param sessionId ID de la session
     * @return Détails de la session
     */
    suspend fun getCallDetails(sessionId: Int): Session {
        return try {
            request("GET", "/api/session/$sessionId") {
                json.decodeFromString<Session>(it)
            }
        } catch (e: SmaRTCException.SessionNotFoundError) {
            throw SmaRTCException.GenericError("Appel introuvable")
        } catch (e: SmaRTCException) {
            throw e
        } catch (e: Exception) {
            throw SmaRTCException.GenericError("Erreur", e)
        }
    }
    
    // ==========================================
    // 🌐 Configuration réseau
    // ==========================================
    
    /**
     * Récupère les serveurs ICE (STUN/TURN)
     * @return Liste des serveurs ICE avec fallback
     */
    suspend fun getIceServers(): List<ICEServer> {
        return try {
            val response = request("GET", "/api/webrtc/ice") {
                json.decodeFromString<ICEServerResponse>(it)
            }
            response.iceServers
        } catch (e: Exception) {
            // Fallback sur Google STUN
            config.stunServers.map { ICEServer(it) }
        }
    }
}
